catroid/src/main/java/org/catrobat/catroid/content/actions/ChangeVariableAction.kt

```diff
@@ -0,0 +1,58 @@
+/*
+ * Catroid: An on-device visual programming system for Android devices
+ * Copyright (C) 2010-2018 The Catrobat Team
+ * (<http://developer.catrobat.org/credits>)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * An additional term exception under section 7 of the GNU Affero
+ * General Public License, version 3, is available at
+ * http://developer.catrobat.org/license_additional_term
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.catrobat.catroid.content.actions
+
+import android.util.Log
+import com.badlogic.gdx.scenes.scene2d.Action
+import org.catrobat.catroid.content.Sprite
+import org.catrobat.catroid.formulaeditor.Formula
+import org.catrobat.catroid.formulaeditor.UserVariable
+
+class ChangeVariableAction : Action() {
+    var sprite: Sprite? = null
+    var changeVariable: Formula? = null
+    var userVariable: UserVariable? = null
+
+    override fun act(delta: Float): Boolean {
+        val originalValue = userVariable?.value ?: return true
+        if (originalValue is String) {
+            return true
+        }
+        var value = changeVariable?.interpretObject(sprite) ?: 0.0
+        if (value is String) {
+            value = try { java.lang.Double.valueOf(value) } catch (e: NumberFormatException) {
+                Log.d(javaClass.simpleName, "Couldn't parse String", e)
+            }
+        }
+        updateUserVariable(originalValue, value)
+        return true
+    }
+
+    private fun updateUserVariable(originalValue: Any, value: Any) {
+        if (originalValue is Double && value is Double) {
+            val original = if (originalValue.isNaN()) 0.0 else originalValue
+            val valueToAdd = if (value.isNaN()) 0.0 else value
+            userVariable?.value = original + valueToAdd
+        }
```

Comment:
https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/take-unless.html

```suggestion
    private fun updateUserVariable(originalValue: Double, value: Double) {
        val original = originalValue.takeUnless { it.isNaN() } ?: 0.0
        val valueToAdd = value.takeUnless { it.isNaN() } ?: 0.0
        userVariable?.value = original + valueToAdd
    }